zproject/backends.py

```diff
@@ -866,6 +884,62 @@ def sync_custom_profile_fields_from_ldap(
         except SyncUserException as e:
             raise ZulipLDAPException(str(e)) from e
 
+    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
+        if not settings.LDAP_SYNC_GROUPS:
+            return
+
+        if user_profile.realm.id != settings.LDAP_GROUP_SYNC_REALM_ID:
+            return
+
+        try:
+            ldap_logger.debug("found groups for user: %s", str(user_profile))
+            ldap_group_name_set = set(ldap_user.group_names)
+
+            member_group_names = [
+                membership.user_group.name
+                for membership in UserGroupMembership.objects.filter(user_profile=user_profile)
+            ]
+
+            ldap_logger.debug("user groups: %s", repr(member_group_names))
+
+            for group_name in ldap_group_name_set:
+                if group_name not in set(settings.LDAP_GROUP_SYNC_LIST):
+                    continue  # this group shall not be synced!
+
+                ldap_logger.debug("group: %s", str(group_name))
+
+                group = UserGroup.objects.get(name=group_name, realm=user_profile.realm)
+
+                # we continue above if group should not be synced and create group if not exists no need to check the group again
+                if not is_user_in_group(
+                    user_group=group, user=user_profile, direct_member_only=True
+                ):
+                    ldap_logger.debug("add %s to %s", user_profile.full_name, str(group_name))
+                    UserGroupMembership.objects.create(
+                        user_group=group,
+                        user_profile=user_profile,
+                    )
+                    do_send_user_group_members_update_event(
+                        event_name="add_members",
+                        user_group=group,
+                        user_ids=[user_profile.id],
+                    )
+
+            for group_name in settings.LDAP_GROUP_SYNC_LIST:
+                group = UserGroup.objects.get(name=group_name, realm=user_profile.realm)
+                if is_user_in_group(user_group=group, user=user_profile, direct_member_only=True):
+                    continue  # we remove here, so skip if in group
+
+                if group_name in member_group_names:
+                    ldap_logger.debug("remove %s from %s", user_profile.full_name, group_name)
+                    UserGroupMembership.objects.get(
+                        user_profile=user_profile,
+                        user_group=group,
+                    ).delete()
```

Comment:
Similar approach can be used here as with `.bulk_create` i noted above. We can get the list of rows to delete and then `.delete()` the whole bulk at once instead of in a loop.